### ISSUE TITLE
Debugger: list blocked asyncs

### ISSUE THREAD

**Issue by Erhannis** [dart-lang/sdk#52732]
Dart SDK version: 2.19.4 (stable) (Tue Mar 7 09:58:46 2023 +0000) on "windows_x64"
Windows
No browser
Android Studio 2022.1.1
(However, issue is applicable to effectively all platforms and up-til-now versions.)

In the past year or so of using Dart (via Flutter, specifically), there is a shortcoming I have repeatedly encountered in the debugging interface: **the debugger can't tell you what asynchronous calls are blocked.**

Having asked on [SO](https://stackoverflow.com/q/73966882/513038) and on Discord, and receiving respectively no response and a very long form of "can't be done", I now post it as a github issue.  Since this is a general usability issue, and not "this particular code doesn't work", code examples don't make as much sense here, but imagine a system wherein a button press demands the completion of a large tree of asynchronous calls.  Network calls are made, database connections are opened and closed, external devices are queried.  If any one of them should (erroneously) block forever, for any reason, the initiating call will likewise block forever.  This is as expected.

However, in most other languages, in my experience (consider Java or Swift, for instance), I would pause the debugger, it would tell me what threads were running and where they were paused, and I could filter, search, or just look through the list for likely culprits until I found the true cause of the holdup.  Generally a matter of five minutes or less.  In Dart, however, when I pause via the debugger, no such list presents itself.  (I understand there's a difference between asynchronous calls and threads, but the principle here is the same.  There exists a number of conceptually concurrent points of execution, and I wish to know what they are.)  As far as I've been able to discern, my best option is "add logging to everything until you see what isn't being called".  In a project spanning dozens of files, or even hundreds, this is a major undertaking - suddenly a five minute task has ballooned into one an hour or more long.  If the trail leads into third party code, adding logs may not even be feasible.

The purpose of the debugger is to help you fix bugs - giving you information about the state of your code, the values of your variables, setting breakpoints, even sometimes evaluating code or modifying variables on-the-fly.  Overall, the Dart debugger does quite a good job of that.  However, "how's execution going" seems like an important category of debugging info that is currently not addressed.  I find this surprising, particularly given Dart's emphasis on asynchronous code.

Can we agree that it would be notably valuable to see, at any time, which lines of your code are blocked?  When might we expect to see such functionality added to the debugger?

**Comment by mkustermann** [comment]
Dart doesn't have real threads, coroutines or fibers, so it doesn't maintain explicit separate stacks for different asynchronous operations. That model is in many ways similar to JavaScript. So one way would be to introduce threads/coroutines/fibers - which could happen in the future.

Though already today, one could partially reconstruct such stacks by do a O(n) heap scan, find all `Future` objects, see who is awaiting on those futures, and who is waiting on those, ... This would yield a DAG of "blocked" suspended functions - where one could print out awaiter stacks.

**Comment by Erhannis** [comment]
Threads/coroutines/fibers may be added: maybe so, though if you can still write code the "old" way, yet not debug its concurrent execution, the "old" way is still short a debugging mechanism.

Scan the heap for Futures: that _sounds_ like the kind of functionality I'd be interested in, though I'm a little unclear of the things it would fail to catch.  Scanning the heap sounds like a thing you'd have to do in the debugger, yeah?  I can't do that from normal code, I expect?

After opening this issue, since I had a problem to solve and don't expect to see any results of this discussion in at least the near future, I wrote extensions for `Future`s and `Future Function`s, adding a wrapping method to insert a registration before execution and a deregistration after.  Then I added it to over 700 `await`ed locations with find/replace.  :P  This DID help me find the probable relevant bugs in my code in short order, but is obviously an abomination and likely computationally expensive.  On registration, a token with a stack trace and timestamp are stored in a tracker singleton, and at any point you can have it print a tree of the traces that have been registered for more than N milliseconds.  In practice then, when your code locks up or something doesn't return, you print the tree (possibly triggered via the debugger) and look for anything suspicious - such as a call to a bluetooth function that should surely have already returned, haha.

The need to manually call `.reg()` on every asynchronous function or future that _might_ be involved in a bug makes it a bad library, and I suspect there's not really a solution to that.  I don't think dart permits injecting code into arbitrary functions at runtime.  The probable performance penalty isn't great either, though I'd accept it in most cases if e.g. additional metadata and processing needed to be added under the hood in order for the debugger to do an equivalent thing invisibly, as long as it could be turned off with e.g. a compilation flag.  There might also need to be some way of marking something "do not track" - I started getting tens of thousands of traces of places that, turns out, were only _sometimes_ completed (think arguments to `Future.any`).  I considered messing with weak references, but at the time it was easier to just not register those places.

TL;DR - I hacked together a rough form of what I wanted, and it did help me fix the bugs I was hunting, but it added a LOT of junk code when implemented on the normal dev side and I expect the debugger/compiler side would be able to do it better in multiple ways.

**Comment by mkustermann** [comment]
> Scanning the heap sounds like a thing you'd have to do in the debugger, yeah? I can't do that from normal code, I expect?

Yes, it's an expensive operation. One would only do it when debugging an app to find bugs. It was my understanding that's exactly your use case: You have an app that's hanging and you'd like to know why.

The benefit of that is that it has absolutely no cost to actual async code when it's running - which is what we want.

**Comment by mkustermann** [comment]
> Sorry - what I meant was, the heap isn't exposed to me, is it?  Scanning
the heap would have to be performed by the debugger itself, something that
has access to dart's internals.  Unless there IS a way for me to scan the
heap?  Whether from code or from some debugging interface I don't know
about, I'd still be interested in hearing.

The implementation would need to scan the keep. Though we'd probably make it available as part of the VM's `vm-service` API (which debuggers are also using). That API is basically a WebSocket protocol anyone can talk to.

/cc @mraleph

**Comment by mkustermann** [comment]
Some years ago I actually did a prototype of this. Now I found the [cl/141603](https://dart-review.googlesource.com/c/sdk/+/141603). Have never pushed to get it productionized and landed.

/cc @aam would you be interested in doing this?

**Comment by aam** [comment]
@a-siva mentioned @derekxu16 is looking at various vm debugger-related issues - cc'ing him on this.
